x/concentrated-liquidity/pool.go

```diff
@@ -418,3 +420,97 @@ func (k Keeper) GetUserUnbondingPositions(ctx sdk.Context, address sdk.AccAddres
 	}
 	return userPositionsWithPeriodLocks, nil
 }
+
+// getPositionIDsByPoolID returns all position IDs for a given pool ID.
+func (k Keeper) getPositionIDsByPoolID(ctx sdk.Context, poolID uint64) ([]uint64, error) {
+	positionIDs, err := osmoutils.GatherValuesFromStorePrefixWithKeyParser(ctx.KVStore(k.storeKey), types.KeyPoolPosition(poolID), parsePositionIDFromPoolLink)
+	if err != nil {
+		return nil, err
+	}
+
+	return positionIDs, nil
+}
+
+// parsePositionIDFromPoolLink parses the position ID from the pool link key.
+func parsePositionIDFromPoolLink(key []byte, _ []byte) (uint64, error) {
+	keys := strings.Split(string(key), types.KeySeparator)
+
+	// Pool ID and position ID.
+	if len(keys) != 2 {
+		return 0, fmt.Errorf("invalid key format when parsing pool/position ID index: %s", string(key))
+	}
+
+	positionID := sdk.BigEndianToUint64([]byte(keys[1]))
+
+	return positionID, nil
+}
+
+// MigrateAccumulatorToScalingFactor multiplies the value of the uptime accumulator, respective position accumulators
+// and tick uptime trackers by the per-unit liquidity scaling factor and overwrites the accumulators with the new values.
+func (k Keeper) MigrateAccumulatorToScalingFactor(ctx sdk.Context, poolId uint64) error {
+	// Get pool-global incentive accumulator
+	uptimeAccums, err := k.GetUptimeAccumulators(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// Get all position IDs for the pool.
+	positionIDs, err := k.getPositionIDsByPoolID(ctx, poolId)
+	if err != nil {
+		return err
+	}
+
+	// For each uptime accimulator, multiply the value by the per-unit liquidity scaling factor
+	// and overwrite the accumulator with the new value.
+	for uptimeIndex, uptimeAccum := range uptimeAccums {
+		value := uptimeAccum.GetValue().MulDecTruncate(perUnitLiqScalingFactor)
+		if err := accum.OverwriteAccumulatorUnsafe(ctx.KVStore(k.storeKey), types.KeyUptimeAccumulator(poolId, uint64(uptimeIndex)), value, uptimeAccum.GetTotalShares()); err != nil {
+			return err
+		}
+
+		// For each position ID, multiply the value by the per-unit liquidity scaling factor
+		// and overwrite the accumulator with the new value.
+		for _, positionID := range positionIDs {
+			positionPrefix := types.KeyPositionId(positionID)
+
+			if !uptimeAccum.HasPosition(string(positionPrefix)) {
+				return fmt.Errorf("position ID %d not found in uptime accumulator %d in pool %d", positionID, uptimeIndex, poolId)
+			}
+
+			positionSnapshot, err := uptimeAccum.GetPosition(string(positionPrefix))
+			if err != nil {
+				return err
+			}
+
+			positionSnapshotValue := positionSnapshot.GetAccumValuePerShare()
+
+			// Multiply the value by the per-unit liquidity scaling factor
+			newValue := positionSnapshotValue.MulDecTruncate(perUnitLiqScalingFactor)
```

Comment:
Is there a mutative method for this?

Reply:
Doesn't look like there is. made issue: https://github.com/cosmos/cosmos-sdk/issues/19389